File: messagebus/src/main/java/com/yahoo/messagebus/SendProxy.java
// Copyright Vespa.ai. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.messagebus;

import com.yahoo.messagebus.network.Network;
import com.yahoo.messagebus.routing.Resender;
import com.yahoo.messagebus.routing.RoutingNode;
import java.util.logging.Level;

import java.util.logging.Logger;

/**
 * This class owns a message that is being sent by message bus. Once a reply is received, the message is attached to it
 * and returned to the application. This also implements the discard policy of {@link RoutingNode}.
 *
 * @author Simon Thoresen Hult
 */
public class SendProxy implements MessageHandler, ReplyHandler {

    private static final Logger log = Logger.getLogger(SendProxy.class.getName());
    private final MessageBus mbus;
    private final Network net;
    private final Resender resender;
    private Message msg = null;
    private boolean logTrace = false;

    /**
     * Constructs a new instance of this class to maintain sending of a single message.
     *
     * @param mbus     The message bus that owns this.
     * @param net      The network layer to transmit through.
     * @param resender The resender to use.
     */
    public SendProxy(MessageBus mbus, Network net, Resender resender) {
        this.mbus = mbus;
        this.net = net;
        this.resender = resender;
    }

    public void handleMessage(Message msg) {
        Trace trace = msg.getTrace();
        if (trace.getLevel() == 0) {
            if (log.isLoggable(Level.FINEST)) {
                trace.setLevel(9);
                logTrace = true;
            } else if (log.isLoggable(Level.FINE)) {
                trace.setLevel(6);
                logTrace = true;
            }
        }
        this.msg = msg;
        RoutingNode root = new RoutingNode(mbus, net, resender, this, msg);
        root.send();
    }

    public void handleReply(Reply reply) {
        if (reply == null) {
            msg.discard();
        } else {
            Trace trace = msg.getTrace();
            if (logTrace) {
                if (reply.hasErrors())
                    log.log(Level.FINE, () -> "Trace for reply with error(s):\n" + reply.getTrace());
                else
                    log.log(Level.FINEST, () -> "Trace for reply:\n" + reply.getTrace());
                Trace empty = new Trace();
                trace.swap(empty);
            } else if (trace.getLevel() > 0) {
                trace.getRoot().addChild(reply.getTrace().getRoot());
                trace.getRoot().normalize();
            }
            reply.swapState(msg);
            reply.setMessage(msg);

            ReplyHandler handler = reply.popHandler();
            handler.handleReply(reply);
        }
    }

}

File: messagebus/src/vespa/messagebus/sendproxy.cpp
// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
#include "sendproxy.h"

#include <vespa/log/log.h>
LOG_SETUP(".sendproxy");

namespace mbus {

SendProxy::SendProxy(MessageBus &mbus, INetwork &net, Resender *resender) :
    _mbus(mbus),
    _net(net),
    _resender(resender),
    _msg(),
    _logTrace(false),
    _root()
{
    // empty
}

void
SendProxy::handleMessage(Message::UP msg)
{
    Trace &trace = msg->getTrace();
    if (trace.getLevel() == 0) {
        if (LOG_WOULD_LOG(spam)) {
            trace.setLevel(9);
            _logTrace = true;
        } else if (LOG_WOULD_LOG(debug)) {
            trace.setLevel(6);
            _logTrace = true;
        }
    }
    _msg = std::move(msg);
    _root.reset(new RoutingNode(_mbus, _net, _resender, *this, *_msg, this));
    _root->send();
}

void
SendProxy::handleDiscard(Context ctx)
{
    (void)ctx;
    _msg->discard();
    delete this;
}

void
SendProxy::handleReply(Reply::UP reply)
{
    Trace &trace = _msg->getTrace();
    if (_logTrace) {
        if (reply->hasErrors()) {
            LOG(debug, "Trace for reply with error(s):\n%s", reply->getTrace().toString().c_str());
        } else if (LOG_WOULD_LOG(spam)) {
            LOG(spam, "Trace for reply:\n%s", reply->getTrace().toString().c_str());
        }
        trace.clear();
    } else if (trace.getLevel() > 0) {
        trace.addChild(reply->steal_trace());
        trace.normalize();
    }
    reply->swapState(*_msg);
    reply->setMessage(std::move(_msg));

    IReplyHandler &handler = reply->getCallStack().pop(*reply);
    handler.handleReply(std::move(reply));

    delete this;
}

} // namespace mbus
